src/wh_client.c

```diff
@@ -119,66 +120,117 @@ int wh_Client_Cleanup(whClient* c)
     return 0;
 }
 
-int wh_Client_EchoRequest(whClient* c, uint16_t size, const void* data)
+int wh_Client_SendRequest(whClient* c,
+        uint16_t group, uint16_t action,
+        uint16_t data_size, const void* data)
 {
-    int rc = 0;
-    whMessageCommLenData msg;
     uint16_t req_id = 0;
-    uint16_t req_type = WOLFHSM_MESSAGE_TYPE_COMM_ECHO;
+    uint16_t kind = WH_MESSAGE_KIND(group, action);
+    int rc = 0;
+
+     rc = wh_CommClient_SendRequest(c->comm,
+                WH_COMM_MAGIC_NATIVE, kind, &req_id,
+                data_size, data);
+    if (rc == 0) {
+        c->last_req_kind = kind;
+        c->last_req_id = req_id;
+    }
+    return rc;
+}
+
+int wh_Client_RecvResponse(whClient *c,
+        uint16_t *out_group, uint16_t *out_action,
+        uint16_t *out_size, void* data)
+{
+    int rc = 0;
+    uint16_t resp_magic = 0;
+    uint16_t resp_kind = 0;
+    uint16_t resp_id = 0;
+    uint16_t resp_size = 0;
+
+    rc = wh_CommClient_RecvResponse(c->comm,
+                &resp_magic, &resp_kind, &resp_id,
+                &resp_size, data);
+    if (rc == 0) {
+        /* Validate response */
+        if (    (resp_magic != WH_COMM_MAGIC_NATIVE) ||
+                (resp_kind != c->last_req_kind) ||
+                (resp_id != c->last_req_id) ){
+            /* Invalid or unexpected message */
+            rc = WH_ERROR_ABORTED;
+        } else {
+            /* Valid and expected message. Set outputs */
+            if (out_group != NULL) {
+                *out_group = WH_MESSAGE_GROUP(resp_kind);
+            }
+            if (out_action != NULL) {
+                *out_action = WH_MESSAGE_ACTION(resp_kind);
+            }
+            if (out_size != NULL) {
+                *out_size = resp_size;
+            }
+        }
+    }
+    return rc;
+}
+
+int wh_Client_EchoRequest(whClient* c, uint16_t size, const void* data)
+{
+    whMessageCommLenData msg = {0};
 
     if (    (c == NULL) ||
             ((size > 0) && (data == NULL)) ){
         return WH_ERROR_BADARGS;
     }
 
-    if (size > sizeof(msg.data)) size = sizeof(msg.data);
+    /* Populate the message */
+    if (size > sizeof(msg.data)) {
```

Comment:
do we want to automagically trim the size on the TX path? or should it be `WH_ERROR_BADARGS`?